test/saveHashTable.m

```diff
@@ -1,37 +1,159 @@
-    fid = fopen(filename,'w+');
-    finally_fclose_fid = onCleanup(@() fclose(fid));
+    % --------------------------------------------------------------------------
+    function hashes = updateHashesFromStatus(hashes, status)
+        % update hashes from the test results in status
+        oldFunctions = fieldnames(hashes);
+        newFunctions = cellfun(@(s) s.function, status, 'UniformOutput', false);
 
-    for iFunc = 1:numel(status)
-        S = status{iFunc};
-        thisFunc = S.function;
-        if isfield(S.hashStage,'found')
-            thisHash = S.hashStage.found;
+        % add hashes from all executed tests
+        for iFunc = 1:numel(status)
+            S = status{iFunc};
+            thisFunc = S.function;
+            thisHash = '';
+            if isfield(S.hashStage,'found')
+                thisHash = S.hashStage.found;
+            elseif S.skip
+                if isfield(hashes, thisFunc)
+                    % Test skipped, but reference hash present in file
+                    % Probably this means that the developer doesn't have access
+                    % to a certain toolbox.
+                    warning('SaveHashTable:CannotUpdateSkippedTest', ...
+                            'Test "%s" was skipped. Cannot update hash!',...
+                            thisFunc);
+                else
+                    % Test skipped and reference hash absent.
+                    % Probably the test is skipped because something is tested
+                    % that relies on HG1/HG2/Octace-specific features and we are
+                    % in the wrong environment for the test.
+                end
+            else
+                warning('SaveHashTable:NoHashFound',...
+                        'No hash found for "%s"!', thisFunc);
+            end
+            if ~isempty(thisHash)
+                hashes.(thisFunc) = thisHash;
+            end
+        end
+
+        % ask what to do with tests for which we have a hash, but no test results
+        removedTests = setdiff(oldFunctions, newFunctions);
+        if ~isempty(removedTests)
+            fprintf(1, 'Some tests in the file were not in the build status.\n');
+        end
+        for iTest = 1:numel(removedTests)
+            thisTest = removedTests{iTest};
+
+            action = askActionToPerformOnRemovedTest(thisTest);
+            switch action
+                case 'remove'
+                    % useful for test that no longer exist
+                    fprintf(1, 'Removed hash for "%s"\n', thisTest);
+                    hashes = rmfield(hashes, thisTest);
+
+                case 'keep'
+                    % useful when not all tests were executed by the tester
+                    fprintf(1, 'Kept hash for "%s"\n', thisTest);
+
+            end
+        end
+    end
+    function action = askActionToPerformOnRemovedTest(testName)
+        % ask which action to carry out on a removed test
+        action = lower(ipp.Results.removedTests);
```

Comment:
That's because most of these functions are [nested functions](http://nl.mathworks.com/help/matlab/matlab_prog/nested-functions.html) (not sub-functions!).

Regarding variables, scoping of a nested function = parameters + variables of parent + own variables.
In Octave there are some things to watch out for (i.e. you can't pass handles to nested functions around), which are acceptable in MATLAB.
Recent versions of MATLAB (2009b or newer, I think) also highlight variables that cross scopes like that (by default in a blue-greenish color).

The same mechanism is used for `FILEFORMAT` by the way.